### PR TITLE
Add example automation scripts.

### PR DESCRIPTION
**Pull request by craigyam**
- Add startup/cleanup and bookinfo demo scripts.
- Add example targets to the Makefile.
- Add HelloWorld test automation changes
- Add scripts to setup kubernetes environment for testing.
- Add make examples to the before_deploy. Also, check the branch
name 'example_automation' as that's the current branch name for
this work.
- Need to install the dependencies (docker-compose, a8ctl etc)
when doing the deploy since make examples needs them.
- Automate the helloworld demo.
- Add traffic-step and traffic-abort and update beta cli binary.
- Tag images with the branch version number and support the cli
recipe-run command.
- Push bookinfo examples to dockerhub.
- Dump debug in bookinfo script after failures.
- Reset the rules on failures to see if that resolves the issues
with the nginx service reporting as down.
- Add additional retries to the remaining testcases.
- Add k8s commands when dump debug info.
- Seeing other failures during some of the other testcases that
will probably be fixed by retries.
- Use kubectl commands for k8s demo runs when dumping debug data.
- Add the start/stop makefile targets to set the version. Fix the
version variable name in the push to dockerhub script.
- Add make targets to stop/start bookinfo.
- Fix version tag for examples images.
- Move push to after_deploy in travis.yaml.
- The examples are running twice as the push script is considered
another deploy.  Move it to the after_deploy section.
Also, add stop/start makefile support for bookinfo.
- Call push script in after_deploy.
- Update push script and move versioned yamls to subdirs.
- Merge in examples updates to change to k8s registry.
- Update with elasticsearch changes.
- Build the cli before running the tests.
- Dockerize before running examples

### CODE REVIEW

**Review by rshriram — CHANGES_REQUESTED** [review]
@craigyam this patch is super huge. Can you break this up into smaller focussed commits that make things easier to review?

There is also an awful lot of duplication in the test scripts between testing and examples/test-scripts folder. Is there any way to deduplicate the stuff?

**Comment by craigyam** [comment]
@rshriram @ijsnellf 
I have broken up the commit into smaller pieces so hopefully that makes reviewing a little easier.  I talked to @gicaceres about reusing code between examples and testing.  He's currently making a number of updates to testing so we thought we'd work together after he's done to see if there is a good way to share common code (in a separate pull request).  Sound good?

File: examples/apps/bookinfo/build-services.sh [hunk 1]
```diff
@@ -16,35 +16,40 @@
 
 set -o errexit
 
+APP_VER=$1
+if [ -z $APP_VER ]; then
+	APP_VER="unknown"
```
Comment (rshriram) [inline examples/apps/bookinfo/build-services.sh:21]: we cannot build with unknown version. We should rather exit

File: .travis.yml [hunk 2]
```diff
@@ -23,6 +24,9 @@ before_install:
         chmod +x docker-compose;
         sudo mv docker-compose /usr/local/bin;
         sudo pip install git+https://github.com/amalgam8/a8ctl@master;
+        wget http://stedolan.github.io/jq/download/linux64/jq;
+        chmod +x ./jq;
+        sudo mv jq /usr/local/bin;
       fi
```
Comment (rshriram) [inline .travis.yml:30]: I think we already have jq in the travis vm. Our existing tests use it IIRC. right?

File: .travis.yml [hunk 3]
```diff
@@ -61,18 +66,9 @@ deploy:
...
-    
-        # Keep artifacts produced during the build
-        skip_cleanup: true
-            
-        # Trigger only when building a version-tagged commit, on the origin repo
-        on:
-            tags: true
-            repo: amalgam8/amalgam8
-            condition: $TRAVIS_TAG =~ v[0-9]+\.[0-9]+\.[0-9]+
+    - ./scripts/push_dockerhub.sh
```
Comment (rshriram) [inline .travis.yml:71]: Why did you remove the check for version-tagged commit ?

File: Makefile [hunk 4]
```diff
@@ -281,6 +282,77 @@ release.examples:
...
+
+examples.buildbookinfo:
+	@echo "--> building examples bookinfo"
+	@examples/apps/bookinfo/build-services.sh $(APP_VER_ABBR)
+
+examples.run: examples.runhello examples.runbookinfo
+
+examples.runhello: build.cli
+	@echo "--> running examples hello world"
+	@examples/test-scripts/docker/test-docker-helloworld.sh $(APP_VER_ABBR) || exit 1
```
Comment (rshriram) [inline Makefile:311]: You don't need `exit 1` in make file.. It should exit by default.

File: examples/k8s-start-helloworld.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+fi
+export APP_VER=$APP_VER
+
+SCRIPTDIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+
+export A8_REGISTRY_URL=http://localhost:31300
+export A8_CONTROLLER_URL=http://localhost:31200
+export GATEWAY_URL=http://localhost:32000
+
+sed -e "s/{APP_VER}/$APP_VER/" $SCRIPTDIR/test-scripts/kubernetes/k8s-helloworld.yaml | kubectl create -f -
```
Comment (rshriram) [inline examples/k8s-start-helloworld.sh:32]: Is this file just for testing or for release? We don't need this file for release.

File: examples/test-scripts/bookinfo.sh [hunk 6]
```diff
@@ -0,0 +1,491 @@
+#!/bin/bash
+#
```
Comment (rshriram) [inline examples/test-scripts/bookinfo.sh:2]: reuse with testing/scripts ?

File: examples/test-scripts/docker/cleanup-helloworld.sh [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+APP_VER=$1
+if [ -z $APP_VER ]; then
+        APP_VER="latest"
+fi
+export APP_VER=$APP_VER
+
+docker-compose -f $SCRIPTDIR/docker-helloworld.yaml kill
+docker-compose -f $SCRIPTDIR/docker-helloworld.yaml rm -f
+
```
Comment (rshriram) [inline examples/test-scripts/docker/cleanup-helloworld.sh:28]: why do we need this file?

File: examples/test-scripts/docker/docker-bookinfo.yaml [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    - elasticsearch
+    - controller
+    - registry
+  container_name: gateway
+  command: /usr/bin/run_filebeat.sh
+
+
+details-v1:
+  image: amalgam8/a8-examples-bookinfo-details-sidecar-v1:${APP_VER}
+  environment:
```
Comment (rshriram) [inline examples/test-scripts/docker/docker-bookinfo.yaml:38]: Our test environment docker file is 99% identical. I suggest we export a bunch of env vars and use that to test actual example apps or test apps, using one single docker file.

File: examples/test-scripts/docker/run-controlplane-docker.sh [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    done
+
+elif [ "$1" == "stop" ]; then
+    echo "Stopping control plane services..."
+    docker-compose -f $SCRIPTDIR/../../docker-controlplane.yaml kill
+    docker-compose -f $SCRIPTDIR/../../docker-controlplane.yaml rm -f
+else
+    echo "usage: $0 start|stop"
+    exit 1
+fi
```
Comment (rshriram) [inline examples/test-scripts/docker/run-controlplane-docker.sh:79]: Kill file. Please see comments on #463

File: examples/test-scripts/docker/test-docker-bookinfo.sh [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+# Run the actual test workload
+$SCRIPTDIR/../bookinfo.sh "docker"
+RC=$?
+
+echo "Docker tests completed with return code: $RC. Cleaning up.."
+$SCRIPTDIR/cleanup-bookinfo.sh $APP_VER
+
+sleep 10
+
+exit $RC
```
Comment (rshriram) [inline examples/test-scripts/docker/test-docker-bookinfo.sh:52]: Kill file. Please see comments on #463

File: examples/test-scripts/gremlin-script/load-productpage.sh [hunk 11]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+# This script runs after the a8ctl recipe-run sets the rules, but before the logs
+# are checked.
+sleep 20
+curl -s -b 'foo=bar;user=jason;x' http://localhost:32000/productpage/productpage > /dev/null
```
Comment (rshriram) [inline examples/test-scripts/gremlin-script/load-productpage.sh:6]: kill. This is a duplicate.

File: examples/test-scripts/helloworld.sh [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+        elif [[ "${temp_var1}" = *"Hello version: version=v2"* ]]; then
+            (( v2_count=v2_count+1 ))
+        fi
+    done
+    echo "    v1 was hit: "$v1_count" times"
+    echo "    v2 was hit: "$v2_count" times"
+    echo ""
+
+    if [ $v1_count -lt 70 ] || [  $v1_count -gt 80 ]; then
+        # Test failed, routing outside 75%/25%, try again
```
Comment (rshriram) [inline examples/test-scripts/helloworld.sh:217]: I am not sure you can rely on such accurate statistics, esp given that choice of versions is random. You might have to run it 1000 times.. Which would take forever.

File: examples/test-scripts/kubernetes/cleanup-bookinfo.sh [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+SCRIPTDIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+
+APP_VER=$1
+if [ -z $APP_VER ]; then
+        APP_VER="latest"
+fi
+
+sed -e "s/{APP_VER}/$APP_VER/" $SCRIPTDIR/k8s-bookinfo.yaml | kubectl delete -f - || echo "Probably already down"
+sleep 2
+$SCRIPTDIR/run-controlplane-kubernetes.sh stop
```
Comment (rshriram) [inline examples/test-scripts/kubernetes/cleanup-bookinfo.sh:27]: Kill file. Please see comments on #463

File: examples/test-scripts/kubernetes/cleanup-helloworld.sh [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+SCRIPTDIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+
+APP_VER=$1
+if [ -z $APP_VER ]; then
+        APP_VER="latest"
+fi
+
+sed -e "s/{APP_VER}/$APP_VER/" $SCRIPTDIR/k8s-helloworld.yaml | kubectl delete -f - || echo "Probably already down"
+sleep 2
+$SCRIPTDIR/run-controlplane-kubernetes.sh stop
```
Comment (rshriram) [inline examples/test-scripts/kubernetes/cleanup-helloworld.sh:27]: kill file

Comment (rshriram) [reply examples/test-scripts/kubernetes/cleanup-helloworld.sh:27]: Move the one liner code into the parent file.

File: examples/test-scripts/kubernetes/install-kubernetes.sh [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        --hostname-override=127.0.0.1 \
+        --address=0.0.0.0 \
+        --api-servers=http://0.0.0.0:8080 \
+        --config=/etc/kubernetes/manifests \
+        --allow-privileged=true --v=2
+
+# Install kubernetes CLI
+type kubectl >/dev/null 2>&1 || install_kubectl
+echo "Waiting for K8S to initialize.."
+sleep 30
```
Comment (rshriram) [inline examples/test-scripts/kubernetes/install-kubernetes.sh:51]: reuse from testing

File: examples/test-scripts/kubernetes/k8s-bookinfo.yaml [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+        - name: A8_DISCOVERY_ADAPTER
+          value: kubernetes
+        - name: A8_CONTROLLER_URL
+          value: http://$(CONTROLLER_SERVICE_HOST):$(CONTROLLER_SERVICE_PORT)
+        - name: A8_CONTROLLER_POLL
+          value: 5s
+        - name: A8_ELASTICSEARCH_SERVER
+          value: $(LOGSERVER_SERVICE_HOST):9200
+        - name: A8_SERVICE
+          value: productpage:version=v1
```
Comment (rshriram) [inline examples/test-scripts/kubernetes/k8s-bookinfo.yaml:312]: same comment as docker file.

File: examples/test-scripts/kubernetes/k8s-helloworld.yaml [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+    spec:
+      containers:
+      - name: helloworld
+        image: amalgam8/a8-examples-helloworld-v2:{APP_VER}
+        imagePullPolicy: IfNotPresent
+        ports:
+        - containerPort: 5000
+        env:
+        - name: A8_SERVICE
+          value: helloworld:version=v2
```
Comment (rshriram) [inline examples/test-scripts/kubernetes/k8s-helloworld.yaml:117]: If this is the only difference in all these files, why can't we use bunch of environment variables ?

File: examples/test-scripts/kubernetes/run-controlplane-kubernetes.sh [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    # done
+
+elif [ "$1" == "stop" ]; then
+    echo "Stopping control plane services..."
+    kubectl delete -f $SCRIPTDIR/../../k8s-controlplane.yaml || echo "Probably already stopped"
+    sleep 3
+else
+    echo "usage: $0 start|stop"
+    exit 1
+fi
```
Comment (rshriram) [inline examples/test-scripts/kubernetes/run-controlplane-kubernetes.sh:75]: kill file.. see #463

File: scripts/push_dockerhub.sh [hunk 19]
```diff
@@ -45,6 +45,25 @@ patch_tag="$major.$minor.$patch"
 minor_tag="$major.$minor"
 major_tag="$major"
 
+echo "Tag: $tag"
+
```
Comment (rshriram) [inline scripts/push_dockerhub.sh:49]: Shouldn't we follow the same convention as sidecar, controller and registry?

**Review by rshriram — CHANGES_REQUESTED** [review]
@craigyam Sorry for the delayed review. There is a lot of duplication between this and the test infrastructure we currently have. IF we take this patch as is, it will get immensely complicated to maintain two sets of scripts that do exactly the same thing.

I suggest reusing the testing scripts in the testing/ folder. Add additional code there to test out the apps in examples/ instead of testing out the apps in testing/apps/. 

Once that is done, the only other change that you would have to tackle is handling JSON (testapps) vs html (existing examples). I think this should be an easy fix.

With regard to helloworld, I suggest doing some simple statistics (50/50) and testing it over 100, and seeing if each version gets atleast 40+ hits.

**Pull request closed without merging**